### PR TITLE
Add support for inline task configuration

### PR DESCRIPTION
**Pull request by ericliuche**
This change adds two new decorators, `@airplane.task` and `@airplane.workflow` to support defining python tasks inline. The decorators wrap a user's function and attaches an `__airplane` object to indicate the function is an Airplane task.

The decorators take in parameters to specify task metadata, e.g. the task's name, slug, schedules, etc. For the task's parameters, the decorators inspect the function's arguments and extract their types. A function 

```
@airplane.task()
def my_func(param: str):
    ...
```
will create a task with a single `shorttext` parameter. The mapping for built in types to airplane types is
```
str -> shorttext
bool -> boolean
int -> integer
float -> float
datetime.date -> date
datetime.datetime -> datetime
```
we also provide additional custom types: `SQL`, `LongText`, `File`, `ConfigVar` that cover the rest of airplane's supported types.

We also add a parameter configuration object, `airplane.ParamConfig` that users can annotate their parameters with to provide additional configuration on their parameter with `typing.Annotated`. e.g. to specify a parameter's name, a user can do
```
@airplane.task()
def my_func(
    param: Annotated[str, airplane.ParamConfig(name="My param name")]
):
    ...
```

Task/parameter descriptions and be provided either through the decorator/ParamConfig or through the function's docstring. The former takes precedence if both are provided.

To support runtime execution of these tasks, the task definition object `TaskDef` provides a `run` function that takes in a dictionary of parameter values from the airplane runtime. It is responsible for filling in None values for optional parameters that don't specify a default as well as coercing types, e.g. converting between a date string and a date object.

### CODE REVIEW

**Review by ameyakhare — CHANGES_REQUESTED** [review]
requesting changes just for feedback rn, will approve when addressed

**Review by ameyakhare — APPROVED** [review]
approving to unblock

File: airplane/exceptions.py [hunk 1]
```diff
@@ -32,3 +35,23 @@ class RunTerminationException(Exception):
...
+    param_name: str
+    prefix: str
+
+    def __str__(self) -> str:
+        return dedent(
+            f"""{self.prefix} for parameter `{self.param_name}` of
+            function `{self.func_name}`.
+
+            Type must be one of (str, int, float, bool, datetime.date, datetime.datetime,
+            airplane.LongText, airplane.File, airplane.ConfigVar, airplane.SQL,
```
[the remark this answers is not in the document]
Comment (ericliuche) [reply airplane/exceptions.py:54]: are you saying we want both `bool` and `airplane.Boolean`?

File: pyproject.toml [hunk 2]
```diff
@@ -12,11 +12,15 @@ packages = [
 ]
 
 [tool.poetry.dependencies]
-python = "^3.7.0"
+python = "^3.9.0"
```
[the remark this answers is not in the document]
Comment (ericliuche) [reply pyproject.toml:15]: airplane's python runtime version is already requiring users to be on 3.10, so i don't believe this should be an issue. we are using a few functions in `typing` that are not available until 3.9

File: .github/CONTRIBUTING.md [hunk 3]
```diff
@@ -11,7 +11,7 @@ $ pyenv install
 Set up a virtual environment and install dependencies with [`poetry`](https://python-poetry.org/docs/) (`brew install poetry`):
 
 ```sh
-$ poetry env use python3.7
+$ poetry env use python3.9
```
[the remark this answers is not in the document]
Comment (ericliuche) [reply .github/CONTRIBUTING.md:14]: same response with 3.9

File: .github/workflows/static-analysis.yml [hunk 4]
```diff
@@ -10,7 +10,7 @@ jobs:
       - uses: actions/checkout@v3
       - uses: actions/setup-python@v3
         with:
-          python-version: 3.7
+          python-version: 3.9
```
[the remark this answers is not in the document]
Comment (ericliuche) [reply .github/workflows/static-analysis.yml:13]: same response with 3.9

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -10,7 +10,7 @@ jobs:
       - uses: actions/checkout@v3
       - uses: actions/setup-python@v3
         with:
-          python-version: 3.7
+          python-version: 3.9
```
[the remark this answers is not in the document]
Comment (ericliuche) [reply .github/workflows/tests.yml:13]: same response with 3.9

File: .pylintrc [hunk 6]
```diff
@@ -17,5 +17,7 @@ disable = missing-module-docstring,
           logging-fstring-interpolation,
           unspecified-encoding,
           too-many-instance-attributes,
-          too-many-branches
+          too-many-branches,
+          too-many-locals,
+          too-many-return-statements
```
[the remark this answers is not in the document]
Comment (ericliuche) [reply .pylintrc]: lol never been useful

File: airplane/config.py [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+
+
+@dataclasses.dataclass
+class ParamConfig:
+    """User configurable parameter configuration."""
+
+    slug: Optional[str] = None
+    name: Optional[str] = None
+    description: Optional[str] = None
+    options: Optional[Union[Sequence[ParamTypes], List[LabeledOption]]] = None
```
[the remark this answers is not in the document]
Comment (ericliuche) [reply airplane/config.py]: ah right right

File: airplane/config.py [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+    regex: Optional[str] = None
+
+
+TaskType = Literal[
```
[the remark this answers is not in the document]
Comment (ericliuche) [reply airplane/config.py]: fixed

File: airplane/config.py [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+    "file",
+    "integer",
+    "float",
+    "date",
+    "datetime",
+    "configvar",
+]
+
+
+RuntimeType = Literal["workflow", "standard"]
```
[the remark this answers is not in the document]
Comment (ericliuche) [reply airplane/config.py]: still need this to enforce the valid annotations for the task definition

File: airplane/config.py [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+    name: str
+    type: RuntimeType
+    entrypoint_func: str
+    description: Optional[str]
+    require_requests: Optional[bool]
+    allow_self_approvals: Optional[bool]
+    timeout: Optional[int]
+    constraints: Optional[Dict[str, str]]
+    resources: Optional[Dict[str, str]]
+    schedules: Optional[Dict[str, Schedule]]
```
[the remark this answers is not in the document]
Comment (ericliuche) [reply airplane/config.py]: fixed

**Pull request merged**